common/task/parallelTaskProcessor.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/uber/cadence/common/metrics"
+)
+
+type (
+	// ParallelTaskProcessorOptions configs PriorityTaskProcessor
+	ParallelTaskProcessorOptions struct {
+		QueueSize   int
+		WorkerCount int
+		RetryPolicy backoff.RetryPolicy
+	}
+
+	parallelTaskProcessorImpl struct {
+		status       int32
+		tasksCh      chan Task
+		shutdownCh   chan struct{}
+		workerWG     sync.WaitGroup
+		logger       log.Logger
+		metricsScope metrics.Scope
+		options      *ParallelTaskProcessorOptions
+	}
+)
+
+var (
+	// ErrTaskProcessorClosed is the error return when submiting task to a stopped processor
+	ErrTaskProcessorClosed = errors.New("task processor has already shutdown")
+)
+
+// NewParallelTaskProcessor creates a new PriorityTaskProcessor
+func NewParallelTaskProcessor(
+	logger log.Logger,
+	metricsScope metrics.Scope,
+	options *ParallelTaskProcessorOptions,
+) Processor {
+	return &parallelTaskProcessorImpl{
+		status:       common.DaemonStatusInitialized,
+		tasksCh:      make(chan Task, options.QueueSize),
+		shutdownCh:   make(chan struct{}),
+		logger:       logger,
+		metricsScope: metricsScope,
+		options:      options,
+	}
+}
+
+func (p *parallelTaskProcessorImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	p.workerWG.Add(p.options.WorkerCount)
+	for i := 0; i < p.options.WorkerCount; i++ {
+		go p.taskWorker()
+	}
+	p.logger.Info("Parallel task processor started.")
+}
+
+func (p *parallelTaskProcessorImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&p.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	close(p.shutdownCh)
+	if success := common.AwaitWaitGroup(&p.workerWG, time.Minute); !success {
+		p.logger.Warn("Parallel task processor timedout on shutdown.")
+	}
+	p.logger.Info("Parallel task processor shutdown.")
+}
+
+func (p *parallelTaskProcessorImpl) Submit(task Task) error {
+	p.metricsScope.IncCounter(metrics.ParallelTaskSubmitRequest)
+	sw := p.metricsScope.StartTimer(metrics.ParallelTaskSubmitLatency)
+	defer sw.Stop()
+
+	select {
+	case p.tasksCh <- task:
+		return nil
+	case <-p.shutdownCh:
+		return ErrTaskProcessorClosed
+	}
+}
+
+func (p *parallelTaskProcessorImpl) taskWorker() {
+	defer p.workerWG.Done()
+
+	for {
+		select {
+		case <-p.shutdownCh:
+			return
+		case task := <-p.tasksCh:
+			p.executeTask(task)
+		}
+	}
+}
+
+func (p *parallelTaskProcessorImpl) executeTask(task Task) {
+	sw := p.metricsScope.StartTimer(metrics.ParallelTaskTaskProcessingLatency)
+	defer sw.Stop()
+
+	op := func() error {
+		if err := task.Execute(); err != nil {
+			return task.HandleErr(err)
+		}
+		return nil
+	}
+
+	isRetryable := func(err error) bool {
+		if p.isStopped() {
+			return false
+		}
+		return task.RetryErr(err)
+	}
+
+	for {
+		if p.isStopped() {
+			// neither ack or nack here
+			return
+		}
+
+		if err := backoff.Retry(op, p.options.RetryPolicy, isRetryable); err != nil {
+			if !task.RetryErr(err) {
```

Comment:
Is this check needed? Can it nack and re-dispatch after retry?

Reply:
Both works. The answer depends on the behavior you want. If you only want to re-dispatch on task-defined nonretryable error then this check is needed. Otherwise, it can be removed. In my opinion, `RetryErr()` should only return `true` when the task should be retried immediately, and the returned value should be honored by the processor.

Reply:
Will update.

Reply:
Done.